overlord/devicestate/handlers_bootconfig.go

```diff
@@ -0,0 +1,79 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"fmt"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+)
+
+func (m *DeviceManager) doUpdateManagedBootConfig(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		return fmt.Errorf("cannot run update boot config task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var seeded bool
+	err := st.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		// do nothing during first boot & seeding
+		return nil
+	}
+	devCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+
+	if devCtx.Model().Grade() == asserts.ModelGradeUnset {
+		// pre UC20 system, do nothing
+		return nil
+	}
+	if devCtx.ForRemodeling() {
+		// TODO:UC20: we may need to update the boot config when snapd
+		// channel is changed during remodel
+		return nil
+	}
+	// TODO:UC20 update recovery boot config
+	updated, err := boot.UpdateManagedBootConfigs(devCtx)
+	if err != nil {
+		return fmt.Errorf("cannot update boot config assets: %v", err)
+	}
+	if updated {
+		t.Logf("updated boot config assets")
+		// boot assets were updated, request a restart now so that the
+		// situation does not end up more complicated if more updates of
+		// boot assets were to be applied
+		st.RequestRestart(state.RestartSystem)
+	}
+
+	t.SetStatus(state.DoneStatus)
```

Comment:
this is a bit confusing but is more an optimisation, we need to survive a reboot in the middle of this anyway?

Reply:
Yes, if we reboot after replacing the boot config asset, but before marking the task as done, the bootloader's update will be called again, but since the new assets are already be in place and CurrentKernelCommandLines contains only the curernt cmdline, `updated` will be false, and nothing will happen.

Reply:
probably good then to add a comment

```
// minimize wasteful redos
```